### PR TITLE
Fixes issue where rope breaks if it changes size

### PR DESCRIPTION
**Pull request by GoodBoyDigital**
- Recreate Types arrays in rope if the size changes
- Fixed a little rope dirty issue too where it was being set to false
  rather than incremented.
- Pass the data to buffer when calling upload as this will resize the
  buffers automatically

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -103,7 +103,7 @@
         "no-unused-labels": 1,
         "no-useless-call": 1,
         "no-useless-concat": 1,
-        "no-useless-escape": 1,
+        "no-useless-escape": 0,
```
Comment (englercj) [inline .eslintrc.json]: Why was this changed?

Comment (englercj) [inline .eslintrc.json]: Maybe, doesn't look like this affects any changes here; so not sure what it is referring to. Might have been left over from when you had changed things in deprecation file?

Comment (englercj) [inline .eslintrc.json]: If we can revert this one line I'm good to merge this.

Comment (GoodBoyDigital) [inline .eslintrc.json]: slick - will do that later tonight

**Comment by englercj** [comment]
Merging this, I'll manually revert the eslint change.
